Add why and load and save for DataFrames (#20)
JuliaDataScience/JuliaDataScience: contents/df_end.md

## Index and Summarize

> ```{=comment}
> summaries
> ```

Reviewer: ```suggestion
```
Do we need this?

> Let's go back to the example data defined above

Reviewer: ```suggestion
Let's go back to the example `grades_2020()` data defined above
```

```jl
sco("""
without_caption_label( # hide
grades_2020()
) # hide
""")
```

> To get a the **vector** for `name` back, we can use

Reviewer: ```suggestion
To retrieve a **vector** for `name`, we can use:
```

Reviewer: Or "recover"

```{=comment}
These two functions cannot be replaced by inline code due to
`df.name` being converted to the same filename as `df[!, :name]` in Books.jl.
I need to fix it.
```

```jl
@sco JDS.names_grades1()
```

or

```jl
@sco JDS.names_grades2()
```

> For the **row**, say the second row, we can use

Reviewer: ```suggestion
For any **row**, say the second row, we can use:
```

```jl
sco("""
df[2, :]
df = DataFrame(df[2, :]) # hide
without_caption_label(df) # hide
""")
```

> or create a function to give us row `i`

Reviewer: ```suggestion
or create a function to give us any row `i` that we want:
```

```jl
@sc JDS.grade_2020(1)
```

```jl
sco("""
without_caption_label( # hide
grade_2020(2)
) # hide
""")
```

> Continuing on this, we can also get only `names` for the first 2 rows:

Reviewer: There is a lot of "this" hanging around

```jl
@sco JDS.grades_indexing()
```

If we assume that all names in the table are unique, we can also write a function to obtain the grade for a person via their `name`.
To do this, we convert the table back to one of Julia's basic data structures which is capable of creating a mappings, namely dictionaries:

```jl
@sc grade_2020("")
```

```jl
sco("""
grade_2020("Bob")
""")
```

> which works because `zip` loops through `df.name` and `df.grade_2020` at the same time like a zipper:

Reviewer: ```suggestion
which works because `zip()` loops through `df.name` and `df.grade_2020` at the same time like a zipper:
```

Author: Is the idea that `f()` denotes a function which can take any parameters a common idea? I could understand that it is confusing for the reader, since one would think that `f` takes no arguments.

How about `zip(...)` and `DataFrame(...)` instead?

Reviewer: We could do that. I want to make a distinction on `variables`, `functions(...)`. `Packages.jl`, `Types` and `Constructors(...)`.
If you define `function(...)` please let me know so I can standardize it in all the sections that I wrote. I've used the `function()` without the `...`

```jl
sco("""
df = grades_2020()
collect(zip(df.name, df.grade_2020))
""")
```

However, converting a DataFrame to a Dictionary is only useful when the elements are unique.

> When that is not the case, it is time to `filter` (@sec:filter).

Reviewer: ```suggestion
When that is not the case, it is time to `filter()` (@sec:filter).
```

## Filter and Subset {#sec:filter}

> Continuing from the earlier mentioned table.

Reviewer: Before was data, now is table. It is confusing. You are calling `grades_2020` by different names...

```jl
sco("""
without_caption_label( # hide
grades_2020()
) # hide
""")
```

We can filter rows by using `filter(f::Function, df)`.

> This function is very similar to the function `filter(f::Function, V::Vector)` from Julia itself.

Reviewer: Say something about multiple dispatch?

Working with a function `f` for filtering can be a bit difficult to use in practice, but it is very powerful.
As a simple example, we can create a function which checks whether it's input equals "Alice":

```jl
@sc JDS.equals_alice("")
```

```jl
sco("""
equals_alice("Bob")
""")
```

```jl
sco("""
equals_alice("Alice")
""")
```

With this function, we can now filter out all the rows for which `name` equals "Alice"

```jl
sco("""
without_caption_label( # hide
filter(:name => equals_alice, grades_2020())
) # hide
""")
```

> Note that this doesn't only work for DataFrames, but also for vectors:

Reviewer: See you are now using the `:`, we really need to standardize a common practice here.

```jl
sco("""
filter(equals_alice, ["Alice", "Bob", "Dave"])
""")
```

> We can make it a bit less verbose by using an anonymous function:

Reviewer: Damn I need to cover anonymous functions. Oooops I forgot!

```jl
sco("""
filter(n -> n == "Alice", ["Alice", "Bob", "Dave"])
""")
```

or, for the table,

```jl
sco("""
without_caption_label( # hide
filter(:name => n -> n == "Alice", grades_2020())
) # hide
""")
```

> This line can be read as "for each element in the column `:name`, let's call this thing `n`, check whether `n` equals Alice".

Reviewer: I will cover Symbols as a variable type somewhere in Julia 101

Reviewer: Also forgot... Oops!

For some people, this is still to verbose.
Luckily, Julia has added a _partial function application_ of `==`.
The details of these words are not important, only that you can use it via

```jl
sco("""
s = "Workaround a bug in books" # hide
without_caption_label( # hide
filter(:name => ==("Alice"), grades_2020())
) # hide
""")
```

To get all the rows which are **not** Alice, `==` can be replaced by `!=` in all previous examples:

```jl
sco("""
without_caption_label( # hide
filter(:name => !=("Alice"), grades_2020())
) # hide
""")
```

> Now, to show why these functions are so powerful, we can come up with a more complex filter.

Reviewer: ```suggestion
Now, to show why anonymous functions are so powerful, we can come up with a more complex filter.
```

Too much this/these/that etc....
What you think?

Author: Agreed. Thanks for the suggestions. I've tried to replace some.
